src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -183,65 +200,95 @@ struct packet {
+    struct ibv_mr                              *grh_mr;
+    uint16_t                                    mcast_lid;
+    union ibv_gid                               mgid;
+    unsigned                                    max_inline;
+    size_t                                      max_eager;
+    int                                         max_per_packet;
+    int                                         pending_send;
+    int                                         pending_recv;
+    struct ibv_mr                              *pp_mr;
+    char                                       *pp_buf;
+    struct pp_packet                           *pp;
+    uint32_t                                    psn;
+    uint32_t                                    last_psn;
+    uint32_t                                    racks_n;
+    uint32_t                                    sacks_n;
+    uint32_t                                    last_acked;
+    uint32_t                                    naks_n;
+    uint32_t                                    child_n;
+    uint32_t                                    parent_n;
+    int                                         buf_n;
+    struct packet                               p2p_pkt[MAX_COMM_POW2];
+    struct packet                               p2p_spkt[MAX_COMM_POW2];
+    ucc_list_link_t                             bpool;
+    ucc_list_link_t                             pending_q;
+    struct mcast_ctx                            mcast;
+    int                                         reliable_in_progress;
+    int                                         recv_drop_packet_in_progress;
+    struct ibv_recv_wr                         *call_rwr;
+    struct ibv_sge                             *call_rsgs;
+    uint64_t                                    timer;
+    int                                         stalled;
+    int                                         comm_id;
+    void                                       *p2p_ctx;
+    ucc_base_lib_t                             *lib;
+    struct sockaddr_in6                         mcast_addr;
+    ucc_rank_t                                  parents[MAX_COMM_POW2];
+    ucc_rank_t                                  children[MAX_COMM_POW2];
+    int                                         nack_requests;
+    int                                         nacks_counter;
+    int                                         n_prep_reliable;
+    int                                         n_mcast_reliable;
+    int                                         wsize;
+    ucc_tl_mlx5_mcast_join_info_t              *group_setup_info;
+    ucc_service_coll_req_t                     *group_setup_info_req;
+    ucc_status_t                               (*bcast_post) (void*, void*, size_t, ucc_rank_t, ucc_service_coll_req_t**);
+    ucc_status_t                               (*allgather_post) (void*, void*, void*, size_t, ucc_service_coll_req_t**);
```

Comment:
I had the same remark on a previous pr but I still fail to understand what is the benefit of having function pointers in this struct instead of just calling functions

Reply:
please look at https://github.com/openucx/ucc/pull/980#discussion_r1624793241